src/tracker/mod.rs

```diff
@@ -0,0 +1,122 @@
+use std::io::{Read, Write};
+use std::net::TcpStream;
+
+use crate::proto::{host_from_url, parse_provided_or_from_env, Client, GetBatchStatus, Track};
+use crate::{BatchStatus, Error, Progress, ProgressUpdate};
+
+/// Used for retrieving and updating information on a job's execution progress
+/// (see [`Progress`] and [`ProgressUpdate`]), as well for retrieving a batch's status
+/// from the Faktory server (see [`BatchStatus`]).
+///
+/// Fetching a job's execution progress:
+/// ```no_run
+/// use faktory::Tracker;
+/// let job_id = String::from("W8qyVle9vXzUWQOf");
+/// let mut tracker = Tracker::connect(None)?;
+/// if let Some(progress) = tracker.get_progress(job_id)? {
+///     if progress.state == "success" {
+///       # /*
+///         ...
+///       # */
+///     }
+/// }
+/// # Ok::<(), faktory::Error>(())
+/// ```
+/// Sending an update on a job's execution progress:
+/// ```no_run
+/// use faktory::{Tracker, ProgressUpdateBuilder};
+/// let jid = String::from("W8qyVle9vXzUWQOf");
+/// let mut tracker = Tracker::connect(None)?;
+/// let progress = ProgressUpdateBuilder::new(&jid)
+///     .desc("Almost done...".to_owned())
+///     .percent(99)
+///     .build();
+/// tracker.set_progress(progress)?;
+/// # Ok::<(), faktory::Error>(())
+///````
+/// Fetching a batch's status:
+/// ```no_run
+/// use faktory::Tracker;
+/// let bid = String::from("W8qyVle9vXzUWQOg");
+/// let mut tracker = Tracker::connect(None)?;
+/// if let Some(status) = tracker.get_batch_status(bid)? {
+///     println!("This batch created at {}", status.created_at);
+/// }
+/// # Ok::<(), faktory::Error>(())
+/// ```
+pub struct Tracker<S: Read + Write> {
```

Comment:
Why have a separate type for this rather than just having these be methods on `Client`?

Reply:
I did not like a `Producer` to be able to set job progress, and I also don't think that the Producer is the only one who can be interested in the job/bath execution progress. Image we have a process running exclusively for tracking purposes: instantiating a `Producer` (not to mention a `Consumer`) for these purposes just feels weird. Hence separation.

I also believe that the `Client` should be responsible for connecting, reconnecting and issuing commands, and should not know anything about setting and getting progress or batch statuses.

That was my reasoning.

Reply:
This makes me wonder if we should make `Client` public, and have these methods be directly on `Client`. What do you think?

Reply:
I've worked with the Go client the other day when porting one of our integration tests. I remember I got stuck for just a little while where I needed to consume the jobs I had previously published. And yes it turned out I could just use the same construct I had used for pushing those jobs.

I see, the idea is pretty straightforward: you instantiate a client and perform _all_ operations with its help. But I guess this just a different approach, and I am not sure I am a fan of it. I also understand I can be biased :smile: 

I've also just checked the API of the faktory python [bindings](https://github.com/cdrx/faktory_worker_python?tab=readme-ov-file#worker-example). They've  exposing:  `client` (our Producer), `server`, and `Worker` (our Consumer). They are implementing it exactly according to the Faktory pieces description that one can find [here](https://github.com/contribsys/faktory_worker_ruby?tab=readme-ov-file#faktory_worker_ruby) for example. They are using a similar approach(with `faktory.work()`) in [node](https://github.com/jbielick/faktory_worker_node?tab=readme-ov-file#processing-jobs) and [C#](https://github.com/tpod/FaktoryWorker?tab=readme-ov-file#worker) bindings. But for me the `Client` vs `Worker` is a little bit confusing. Also notice how this is _different_ from the Go client API, where there is simply `Client`.

I've worked previously with [Kafka.js](https://kafka.js.org/docs/getting-started) lib where the concept of `producer` and `consumer`  (which is kind of projected from `Kafka` to the bindings) just felt so natural to me.  In most of the cases, you are launching a process to either produce, or consume, so you only need role-specific methods.

Reply:
Ah, so, to be clear, what I'm proposing is that we have a `Client` type, but that you _cannot call any methods on it_ (except maybe the tracker methods). Then we have a `Producer` and `Consumer` type, both of which wrap a `Client` and adds additional methods.

Reply:
I see. Removed the tracker wrapper, exposed `connect` and `connect_with` on the `Client` directly + those `get_batch_status`, `get_progress` , and `set_progress`. 

I also refactored `is_producer` of the `ClientOptions` to be `is_worker` which solved my previous problem, where I did not want to use `new_producer` on the `Client` for the `Tracker`. 

We only want some special configuration for a consumer (worker in Faktory lingo). So `is_worker` will be `false` by default, and only have `true` when this client is being instantiated for `Consumer` purposes.